### PR TITLE
Support for lowering of complex types in 'tfl.reshape' op

### PR DESCRIPTION
**Pull request by rafaelubalmw**

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @rafaelubalmw Can you please sign CLA. Thank you!

**Comment by rafaelubalmw** [comment]
@gbaned Looks like my GitHub user was just successfully added to the MathWorks Google Group associated with an already signed corporate CLA. Happy to address any additional comments. Thanks!

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 1]
```diff
@@ -2145,8 +2146,8 @@ LogicalResult ConvertTFLReshapeOp::matchAndRewrite(
   auto tfl_reshape_op = cast<TFL::ReshapeOp>(op);
 
   Value shape = tfl_reshape_op.getShape();
-  ShapedType shape_type = dyn_cast<ShapedType>(shape.getType());
-  ShapedType output_type = dyn_cast<ShapedType>(tfl_reshape_op.getType());
+  ShapedType shape_type = shape.getType().dyn_cast<ShapedType>();
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:2149]: Note: this is switching to deprecated form (https://mlir.llvm.org/deprecation/)

File: WORKSPACE [hunk 2]
```diff
@@ -86,3 +86,18 @@ tf_workspace1()
 load("@//tensorflow:workspace0.bzl", "tf_workspace0")
 
 tf_workspace0()
+
+# Hedron's Compile Commands Extractor for Bazel
```
Comment (jpienaar) [inline WORKSPACE:90]: Why is this needed?

Comment (rafaelubalmw) [reply WORKSPACE:90]: This allows us to generate file `compile_commands.json` in the root directly by running command `bazelisk run tensorflow/compiler/mlir:tf_opt_compile_commands -- --config=avx2_linux --config=dbg`. Useful for LSP.

I'm aware this change is unrelated with the proposed feature. If you'd like I can submit this as a separate pull request, or remove it altogether if Tensorflow already has a way to integrate with LSP that I'm not aware of. Let me know how you'd like me to proceed.

**Comment by rafaelubalmw** [comment]
@jpienaar @eric-k256 Just pinging on this PR. Thanks!

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 3]
```diff
@@ -2168,7 +2169,7 @@ LogicalResult ConvertTFLReshapeOp::matchAndRewrite(
   // Extract the dynamically shaped values for each dimension.
   SmallVector<Value> shape_vals;
   shape_vals.reserve(rank);
-  auto shape_ty = dyn_cast<ShapedType>(shape.getType());
+  auto shape_ty = shape.getType().dyn_cast<ShapedType>();
```
Comment (eric-k256) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:2172]: I think this should be a cast instead of dyn_cast. The problem is in the original code, but if we're changing the line, we can fix this at the same time.

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 4]
```diff
@@ -2180,13 +2181,40 @@ LogicalResult ConvertTFLReshapeOp::matchAndRewrite(
...
   }
 
+  auto input = tfl_reshape_op.getInput();
+  auto reshape_type = output_type;
+  if (input.getType().getElementType().isa<ComplexType>()) {
+    input = llvm::cast<TypedValue<TensorType>>(
+      tosa::lowerComplexTensor(rewriter, op->getLoc(), input));
+
+    auto constant_type = RankedTensorType::get({},rewriter.getIntegerType(32));
+	// auto constant_attr = DenseI32ArrayAttr::get(rewriter.getContext(), {2});
```
Comment (eric-k256) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:2191]: Leftover code that can be removed?

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 5]
```diff
@@ -31,6 +31,7 @@ limitations under the License.
 #include <unordered_set>
 
 #include "llvm/ADT/ArrayRef.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
```
Comment (eric-k256) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:34]: I don't see any references to arith below, is this needed?
